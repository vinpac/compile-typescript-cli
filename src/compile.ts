import chalk from 'chalk'
import { ChildProcess, spawn } from 'child_process'
import clearConsole from 'clear-console'
import * as path from 'path'
import webpack from 'webpack'
import createWebpackConfig from './create-webpack-config'

export interface CompileOptions {
  entry: string
  output: string
  verbose?: boolean
  execCommand?: string
  dev: boolean
  watch?: boolean
  run?: boolean
  checkTypes?: boolean
  quiet?: boolean
  configPath: string
  onCompile?: () => any
}

export default function compile(options: CompileOptions) {
  const webpackConfig = createWebpackConfig(options)

  let outputDir: string | undefined
  let outputFile = 'index.js'

  if (options.output.endsWith('/')) {
    outputDir = options.output
  } else {
    outputDir = path.dirname(options.output)
    outputFile = path.basename(options.output)
  }

  if (!options.entry) {
    throw new Error("'entry' is a required parameter")
  }

  if (!options.output) {
    throw new Error("'output' is a required parameter")
  }

  const compiler = webpack(webpackConfig)
  let childProcesses: ChildProcess[] = []

  const onChildNodeProcessExit = () => {
    console.info(chalk.red('> Process finished'))
  }

  const onFinish = (elapsedTime: number) => {
    console.info(chalk.green(`> Successfully compiled in ${elapsedTime}ms`))

    if (options.run) {
      const runChild = spawn('node', [path.join(outputDir!, outputFile)], {
        env: Object.assign(
          { NODE_ENV: options.dev ? 'development' : 'production' },
          process.env,
        ),
        stdio: options.quiet ? 'ignore' : 'inherit',
      })
      runChild.once('exit', () => {
        childProcesses = childProcesses.filter(
          childProcess => childProcess !== runChild,
        )
        runChild.removeListener('exit', onChildNodeProcessExit)
        runChild.kill('SIGTERM')

        onChildNodeProcessExit()

        if (
          childProcesses.length === 0 &&
          !options.watch &&
          options.onCompile
        ) {
          options.onCompile()
        }
      })

      childProcesses.push(runChild)
    }

    if (options.execCommand) {
      console.info(chalk.blue('> Running exec command'))
      const [command, ...args] = options.execCommand.split(' ')
      const execChild = spawn(command, args, {
        env: Object.assign(
          { NODE_ENV: options.dev ? 'development' : 'production' },
          process.env,
        ),
        stdio: options.quiet ? 'ignore' : 'inherit',
      })
      execChild.once('exit', onChildNodeProcessExit)
      childProcesses.push(execChild)
    }

    if (
      !options.run &&
      !options.execCommand &&
      !options.watch &&
      options.onCompile
    ) {
      options.onCompile()
    }
  }

  const onCompilationFinish: webpack.ICompiler.Handler = (error, stats) => {
    // Kill current childs
    if (childProcesses.length) {
      childProcesses.forEach(childProcess => {
        childProcess.removeListener('exit', onChildNodeProcessExit)
        childProcess.kill('SIGTERM')

        return false
      })
      childProcesses = []
    }

    if (error || stats.hasErrors()) {
      const failedMessage = chalk.red(
        `> Compilation failed${
          stats && stats.endTime && stats.startTime
            ? ` after ${stats.endTime - stats.startTime}ms`
            : ''
        }`,
      )

      if (error) {
        console.info(failedMessage)
        console.error(error)
        return
      }

      console.info(failedMessage)
      console.info(
        stats.toString({
          assets: options.verbose,
          builtAt: options.verbose,
          entrypoints: options.verbose,
          timings: false,
          cached: options.verbose,
          cachedAssets: options.verbose,
          chunks: options.verbose,
          chunkModules: options.verbose,
          colors: true,
          hash: options.verbose,
          modules: options.verbose,
          reasons: options.dev,
          version: options.verbose,
        }),
      )
      return
    }

    if (!options.watch) {
      onFinish(
        stats.endTime && stats.startTime ? stats.endTime - stats.startTime : 0,
      )
    }
  }

  if (options.checkTypes) {
    // @ts-ignore
    compiler.hooks.forkTsCheckerDone.tap(
      'should',
      (_: any, __: any, elapsed: number) => {
        clearConsole()
        setTimeout(() => onFinish(Math.round(elapsed / 1000000)), 1)
      },
    )
  }

  if (options.watch) {
    compiler.watch({}, onCompilationFinish)
  } else {
    compiler.run(onCompilationFinish)
  }
}
