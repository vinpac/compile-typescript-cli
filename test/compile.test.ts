import { ChildProcess } from 'child_process'
import * as fs from 'fs'
import path from 'path'
import { promisify } from 'util'
import compile, { CompileOptions } from '../src/compile'

const readFile = promisify(fs.readFile)
const fixturesDir = path.resolve('test', '__fixtures__')
const distDir = path.resolve(fixturesDir, 'dist')

const fixture = (filename: string) => path.resolve(fixturesDir, filename)
const compileAsync = (filename: string, options?: Partial<CompileOptions>) => {
  const output = path.resolve(distDir, filename.replace(/\.tsx?$/g, '.js'))

  return new Promise((resolve, reject) =>
    compile({
      dev: true,
      output,
      tsConfigPath: path.resolve(fixturesDir, 'tsconfig.json'),
      entry: fixture(filename),
      quiet: true,
      configPath: undefined,
      run: createRunFn(resolve, reject),
      ...options,
      onCompile: error => (error ? reject(error) : resolve()),
    }),
  ).then(() => readFile(output, 'utf8'))
}

const createRunFn = (
  resolve: (payload: any) => void,
  reject: (error: any) => void,
  outputExpected?: string[],
) => {
  return (child: ChildProcess) => {
    let outputIndex = 0
    if (outputExpected) {
      child.stdout!.on('data', data => {
        expect(String(data)).toBe(outputExpected[outputIndex])
        outputIndex += 1
      })
    }
    child.on('close', code => {
      if (outputExpected && outputIndex !== outputExpected.length) {
        reject(
          new Error(`Expected ${outputExpected.length} logs before exiting`),
        )
        return
      }

      if (code === 1) {
        reject('Exited with code 1')
      } else {
        resolve('Exited')
      }
    })
  }
}

describe('Compile', () => {
  it('should compile a file without problems', async () => {
    const outputPath = path.resolve(distDir, 'basics.js')
    await compileAsync('basics.ts')
    const output = await readFile(outputPath, 'utf8')

    expect(output).toMatchSnapshot('basics')
  })

  it('should use detect baseUrl and add TsconfigPathsPlugin', async () => {
    const output = await compileAsync('paths.ts')
    expect(output).toMatchSnapshot('paths')
  })

  it('should compile into a executable jasvascript file', async () => {
    await new Promise((resolve, reject) => {
      compileAsync('paths.ts', {
        run: createRunFn(resolve, reject, ['Basics\n', 'foo =  foo\n']),
      })
    })
  })

  it("should throw an error if tsconfig doesn't exist", async () => {
    try {
      await compileAsync('paths.ts', {
        tsConfigPath: '__',
      })
    } catch (error) {
      expect(error.message).toBe(
        "Unable to find a valid configuration JSON for typescript at '__'",
      )
      return
    }

    throw Error('Compiled without problems')
  })

  it("should throw an error if the entry file doesn't compile", async () => {
    try {
      await compileAsync('broken-compilation.ts', { silent: true })
    } catch (error) {
      expect(error.message).toMatchSnapshot()
      return
    }

    throw Error('Compiled without problems')
  })

  it('should use tncc.config', async () => {
    await compileAsync('with-react.tsx', {
      configPath: fixture('tncc.config.js'),
    })
  })

  it('should work without output', async () => {
    await compileAsync('with-react.tsx', {
      configPath: fixture('tncc.config.js'),
      output: undefined,
    })
  })
})
