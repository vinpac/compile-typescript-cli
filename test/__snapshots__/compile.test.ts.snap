// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compile should compile a file without problems: basics 1`] = `
"/******/ (function(modules) { // webpackBootstrap
/******/ 	// The module cache
/******/ 	var installedModules = {};
/******/
/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {
/******/
/******/ 		// Check if module is in cache
/******/ 		if(installedModules[moduleId]) {
/******/ 			return installedModules[moduleId].exports;
/******/ 		}
/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = installedModules[moduleId] = {
/******/ 			i: moduleId,
/******/ 			l: false,
/******/ 			exports: {}
/******/ 		};
/******/
/******/ 		// Execute the module function
/******/ 		modules[moduleId].call(module.exports, module, module.exports, __webpack_require__);
/******/
/******/ 		// Flag the module as loaded
/******/ 		module.l = true;
/******/
/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}
/******/
/******/
/******/ 	// expose the modules object (__webpack_modules__)
/******/ 	__webpack_require__.m = modules;
/******/
/******/ 	// expose the module cache
/******/ 	__webpack_require__.c = installedModules;
/******/
/******/ 	// define getter function for harmony exports
/******/ 	__webpack_require__.d = function(exports, name, getter) {
/******/ 		if(!__webpack_require__.o(exports, name)) {
/******/ 			Object.defineProperty(exports, name, { enumerable: true, get: getter });
/******/ 		}
/******/ 	};
/******/
/******/ 	// define __esModule on exports
/******/ 	__webpack_require__.r = function(exports) {
/******/ 		if(typeof Symbol !== 'undefined' && Symbol.toStringTag) {
/******/ 			Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
/******/ 		}
/******/ 		Object.defineProperty(exports, '__esModule', { value: true });
/******/ 	};
/******/
/******/ 	// create a fake namespace object
/******/ 	// mode & 1: value is a module id, require it
/******/ 	// mode & 2: merge all properties of value into the ns
/******/ 	// mode & 4: return value when already ns object
/******/ 	// mode & 8|1: behave like require
/******/ 	__webpack_require__.t = function(value, mode) {
/******/ 		if(mode & 1) value = __webpack_require__(value);
/******/ 		if(mode & 8) return value;
/******/ 		if((mode & 4) && typeof value === 'object' && value && value.__esModule) return value;
/******/ 		var ns = Object.create(null);
/******/ 		__webpack_require__.r(ns);
/******/ 		Object.defineProperty(ns, 'default', { enumerable: true, value: value });
/******/ 		if(mode & 2 && typeof value != 'string') for(var key in value) __webpack_require__.d(ns, key, function(key) { return value[key]; }.bind(null, key));
/******/ 		return ns;
/******/ 	};
/******/
/******/ 	// getDefaultExport function for compatibility with non-harmony modules
/******/ 	__webpack_require__.n = function(module) {
/******/ 		var getter = module && module.__esModule ?
/******/ 			function getDefault() { return module['default']; } :
/******/ 			function getModuleExports() { return module; };
/******/ 		__webpack_require__.d(getter, 'a', getter);
/******/ 		return getter;
/******/ 	};
/******/
/******/ 	// Object.prototype.hasOwnProperty.call
/******/ 	__webpack_require__.o = function(object, property) { return Object.prototype.hasOwnProperty.call(object, property); };
/******/
/******/ 	// __webpack_public_path__
/******/ 	__webpack_require__.p = \\"\\";
/******/
/******/
/******/ 	// Load entry module and return exports
/******/ 	return __webpack_require__(__webpack_require__.s = \\"./test/__fixtures__/basics.ts\\");
/******/ })
/************************************************************************/
/******/ ({

/***/ \\"./test/__fixtures__/basics.ts\\":
/*!*************************************!*\\\\
  !*** ./test/__fixtures__/basics.ts ***!
  \\\\*************************************/
/*! exports provided: foo */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

\\"use strict\\";
eval(\\"__webpack_require__.r(__webpack_exports__);\\\\n/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, \\\\\\"foo\\\\\\", function() { return foo; });\\\\nvar foo = 'foo';\\\\n// tslint:disable-next-line:no-console\\\\nconsole.log('Basics');\\\\n//# sourceURL=[module]\\\\n//# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiLi90ZXN0L19fZml4dHVyZXNfXy9iYXNpY3MudHMuanMiLCJzb3VyY2VzIjpbIi9Vc2Vycy9tdXJkb2NrL2xhYi9vcGVuLXNvdXJjZS90bmNjL3Rlc3QvX19maXh0dXJlc19fL2Jhc2ljcy50cyJdLCJzb3VyY2VzQ29udGVudCI6WyJleHBvcnQgdmFyIGZvbyA9ICdmb28nO1xuLy8gdHNsaW50OmRpc2FibGUtbmV4dC1saW5lOm5vLWNvbnNvbGVcbmNvbnNvbGUubG9nKCdCYXNpY3MnKTtcbiJdLCJtYXBwaW5ncyI6IkFBQUE7QUFBQTtBQUFBO0FBQ0E7QUFDQTsiLCJzb3VyY2VSb290IjoiIn0=\\\\n//# sourceURL=webpack-internal:///./test/__fixtures__/basics.ts\\\\n\\");

/***/ })

/******/ });"
`;

exports[`Compile should throw an error if the entry file doesn't compile 1`] = `
"[90m[tsl] [39m[1m[31mERROR[39m[22m[1m[31m in [39m[22m[1m[36m/Users/murdock/lab/open-source/tncc/test/__fixtures__/broken-compilation.ts(7,13)[39m[22m
[1m[31m      TS1005: ';' expected.[39m[22m"
`;

exports[`Compile should use detect baseUrl and add TsconfigPathsPlugin: paths 1`] = `
"/******/ (function(modules) { // webpackBootstrap
/******/ 	// The module cache
/******/ 	var installedModules = {};
/******/
/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {
/******/
/******/ 		// Check if module is in cache
/******/ 		if(installedModules[moduleId]) {
/******/ 			return installedModules[moduleId].exports;
/******/ 		}
/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = installedModules[moduleId] = {
/******/ 			i: moduleId,
/******/ 			l: false,
/******/ 			exports: {}
/******/ 		};
/******/
/******/ 		// Execute the module function
/******/ 		modules[moduleId].call(module.exports, module, module.exports, __webpack_require__);
/******/
/******/ 		// Flag the module as loaded
/******/ 		module.l = true;
/******/
/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}
/******/
/******/
/******/ 	// expose the modules object (__webpack_modules__)
/******/ 	__webpack_require__.m = modules;
/******/
/******/ 	// expose the module cache
/******/ 	__webpack_require__.c = installedModules;
/******/
/******/ 	// define getter function for harmony exports
/******/ 	__webpack_require__.d = function(exports, name, getter) {
/******/ 		if(!__webpack_require__.o(exports, name)) {
/******/ 			Object.defineProperty(exports, name, { enumerable: true, get: getter });
/******/ 		}
/******/ 	};
/******/
/******/ 	// define __esModule on exports
/******/ 	__webpack_require__.r = function(exports) {
/******/ 		if(typeof Symbol !== 'undefined' && Symbol.toStringTag) {
/******/ 			Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
/******/ 		}
/******/ 		Object.defineProperty(exports, '__esModule', { value: true });
/******/ 	};
/******/
/******/ 	// create a fake namespace object
/******/ 	// mode & 1: value is a module id, require it
/******/ 	// mode & 2: merge all properties of value into the ns
/******/ 	// mode & 4: return value when already ns object
/******/ 	// mode & 8|1: behave like require
/******/ 	__webpack_require__.t = function(value, mode) {
/******/ 		if(mode & 1) value = __webpack_require__(value);
/******/ 		if(mode & 8) return value;
/******/ 		if((mode & 4) && typeof value === 'object' && value && value.__esModule) return value;
/******/ 		var ns = Object.create(null);
/******/ 		__webpack_require__.r(ns);
/******/ 		Object.defineProperty(ns, 'default', { enumerable: true, value: value });
/******/ 		if(mode & 2 && typeof value != 'string') for(var key in value) __webpack_require__.d(ns, key, function(key) { return value[key]; }.bind(null, key));
/******/ 		return ns;
/******/ 	};
/******/
/******/ 	// getDefaultExport function for compatibility with non-harmony modules
/******/ 	__webpack_require__.n = function(module) {
/******/ 		var getter = module && module.__esModule ?
/******/ 			function getDefault() { return module['default']; } :
/******/ 			function getModuleExports() { return module; };
/******/ 		__webpack_require__.d(getter, 'a', getter);
/******/ 		return getter;
/******/ 	};
/******/
/******/ 	// Object.prototype.hasOwnProperty.call
/******/ 	__webpack_require__.o = function(object, property) { return Object.prototype.hasOwnProperty.call(object, property); };
/******/
/******/ 	// __webpack_public_path__
/******/ 	__webpack_require__.p = \\"\\";
/******/
/******/
/******/ 	// Load entry module and return exports
/******/ 	return __webpack_require__(__webpack_require__.s = \\"./test/__fixtures__/paths.ts\\");
/******/ })
/************************************************************************/
/******/ ({

/***/ \\"./test/__fixtures__/basics.ts\\":
/*!*************************************!*\\\\
  !*** ./test/__fixtures__/basics.ts ***!
  \\\\*************************************/
/*! exports provided: foo */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

\\"use strict\\";
eval(\\"__webpack_require__.r(__webpack_exports__);\\\\n/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, \\\\\\"foo\\\\\\", function() { return foo; });\\\\nvar foo = 'foo';\\\\n// tslint:disable-next-line:no-console\\\\nconsole.log('Basics');\\\\n//# sourceURL=[module]\\\\n//# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiLi90ZXN0L19fZml4dHVyZXNfXy9iYXNpY3MudHMuanMiLCJzb3VyY2VzIjpbIi9Vc2Vycy9tdXJkb2NrL2xhYi9vcGVuLXNvdXJjZS90bmNjL3Rlc3QvX19maXh0dXJlc19fL2Jhc2ljcy50cyJdLCJzb3VyY2VzQ29udGVudCI6WyJleHBvcnQgdmFyIGZvbyA9ICdmb28nO1xuLy8gdHNsaW50OmRpc2FibGUtbmV4dC1saW5lOm5vLWNvbnNvbGVcbmNvbnNvbGUubG9nKCdCYXNpY3MnKTtcbiJdLCJtYXBwaW5ncyI6IkFBQUE7QUFBQTtBQUFBO0FBQ0E7QUFDQTsiLCJzb3VyY2VSb290IjoiIn0=\\\\n//# sourceURL=webpack-internal:///./test/__fixtures__/basics.ts\\\\n\\");

/***/ }),

/***/ \\"./test/__fixtures__/paths.ts\\":
/*!************************************!*\\\\
  !*** ./test/__fixtures__/paths.ts ***!
  \\\\************************************/
/*! no exports provided */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

\\"use strict\\";
eval(\\"__webpack_require__.r(__webpack_exports__);\\\\n/* harmony import */ var _basics__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! ~/basics */ \\\\\\"./test/__fixtures__/basics.ts\\\\\\");\\\\n// @ts-ignore\\\\n\\\\n// tslint:disable-next-line:no-console\\\\nconsole.log('foo = ', _basics__WEBPACK_IMPORTED_MODULE_0__[\\\\\\"foo\\\\\\"]);\\\\n//# sourceURL=[module]\\\\n//# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiLi90ZXN0L19fZml4dHVyZXNfXy9wYXRocy50cy5qcyIsInNvdXJjZXMiOlsiL1VzZXJzL211cmRvY2svbGFiL29wZW4tc291cmNlL3RuY2MvdGVzdC9fX2ZpeHR1cmVzX18vcGF0aHMudHMiXSwic291cmNlc0NvbnRlbnQiOlsiLy8gQHRzLWlnbm9yZVxuaW1wb3J0IHsgZm9vIH0gZnJvbSAnfi9iYXNpY3MnO1xuLy8gdHNsaW50OmRpc2FibGUtbmV4dC1saW5lOm5vLWNvbnNvbGVcbmNvbnNvbGUubG9nKCdmb28gPSAnLCBmb28pO1xuIl0sIm1hcHBpbmdzIjoiQUFBQTtBQUFBO0FBQUE7QUFDQTtBQUNBO0FBQ0E7Iiwic291cmNlUm9vdCI6IiJ9\\\\n//# sourceURL=webpack-internal:///./test/__fixtures__/paths.ts\\\\n\\");

/***/ })

/******/ });"
`;
